### ISSUE TITLE
Implement nr option?

### ISSUE THREAD

**Issue by hemberger** [MechanicalSoup/MechanicalSoup#130]
In MechanicalSoup, almost all of our methods that find some element return or operate on the first one that is found (`Form.choose_submit` appears to be the recently-modified exception). If, for example, you have a page with two `<form>...</form>` with no unique attributes, then you cannot select the second form with `StatefulBrowser.select_form`.

This problem is solved in Mechanize by giving these methods an `nr=1` argument, which selects the _nr_-th match, and the first by default. We could certainly add this to many of the methods and retain the current default behavior entirely. I think this alone would be a worthwhile enhancement.

One related (but perhaps separate) issue is the lack of feedback when you omit the `nr` option and get an element that you don't expect, likely resulting in an indecipherable error, or worse, silently incorrect behavior. In my ideal world, you'd specify `nr` when you know you want the _nr_-th match, and if `nr` is omitted you demand that there is only one match. This seems like the safest option, but may be too draconian. Perhaps it would be suitable for a "debug" or "strict" mode?

**Comment by moy** [comment]
I agree it would be nice to have.

I'm not sure what's the best option between being defensive (error out when there are more than 1) or just pick the first. I like defensive code, but in our case it's possible that the developer wrote the script at a time when there was just one form, and later the site adds another form to the page. Then you may as well get lucky and try the first (if it's the wrong one, you'll almost certainly get an error while filling-in the form). Probably the ideal behavior would be to have both indeed.

**Comment by hemberger** [comment]
This seems like a good enhancement to work on for the version after 0.9.0 (1.0.0?). Do you like to use GitHub's milestones?

**Comment by moy** [comment]
Why not. I have mixed feelings about milestones for community driven projects though: on the one hand they help sorting the list of issues, but on the other hand in practice milestones assignments are very often empty promises.

**Comment by hemberger** [comment]
I ran into a possible syntactical ambiguity that will affect backwards compatibility. The current interface for `StatefulBrowser.select_form` is:
```python
def select_form(self, selector="form", *args, **kwargs):
    ...
```
Now, if I want to add an `nr=1` argument,
```python
def select_form(self, selector="form", nr=1, *args, **kwargs):
    ...
```
I don't see how I can do this without breaking previously valid calls, because any argument that was previously part of `*args` will now get slurped into `nr`.

For example, `select_form("form", None, 1)` would previously give `args=(None, 1)` but now would give `nr=None, args=(1,)`.

Additional reading:
https://stackoverflow.com/questions/15301999/python-2-x-default-arguments-with-args-and-kwargs
https://stackoverflow.com/questions/9872824/calling-a-python-function-with-args-kwargs-and-optional-default-arguments

--------------------------

Now, we should also consider the current interface for `BeautifulSoup.select`:
```python
BeautifulSoup.select(self, selector, _candidate_generator=None, limit=None)
```
which we call from `StatefulBrowser.select_form` using:
```python
found_forms = self.__current_page.select(selector, *args, **kwargs)
```
It seems unlikely that anyone would be using `_candidate_generator`, but it's possible that someone was using `limit` (specified as a kwarg). If we _cannot avoid_ breaking backwards compatibility for this function interface, then I would suggest to avoid variable arguments entirely with the following:

```python
def select_form(self, selector="form", nr=1):
    found_forms = self.__current_page.select(selector, limit=nr)
    ...
```
This has two benefits:
1. It makes the backwards compatibility errors more transparent:
    - `_candidate_generator` specified as a kwarg -> `TypeError: select_form() got an unexpected keyword '_candidate_generator'`
    - `limit` specified as a kwarg -> `SyntaxError: keyword argument repeated` (arguably the most important case)
    - `_candidate_generator` and `limit` both specified as args -> `TypeError: select_form() takes at most 3 arguments (4 given)`
    - `_candidate_generator` alone specified as an arg -> this cannot be disambiguated, but its value is either going to be `None` or "not an int", so it will likely be an obviously bad value for `nr`.
2. It automatically adds in the efficiency gain of specifying the `limit` option of `BeautifulSoup.select`, which doesn't make sense to specify independently of an `nr` option.

**Comment by moy** [comment]
IIRC I'm the one who forwarded `*args` and `**kwargs`, and if so I did it just because I did not know which exact arguments BeautifulSoup's `select` was taking, so I forwarded for maximal generality. `_candiate_generator` is hardly documented (undocumented?), and the fact that it starts with `_` suggests that it's an internal detail that users shouldn't use. People calling `_candidate_error` through MechanicalSoup would get what they deserve if we break their code ;-).

Since `limit` comes after `_candidate_genrator`, it cannot be passed through `*args`, but only through `*kwargs`. I'd say we can safely trash `*args`, and I do not care deeply if we trash `*kwargs` as you suggest, too.

I think there's a minor error in your explanations: "limit specified as a kwarg" -> there's no kwargs anymore so the error is that there's no such argument. If you keep forwarding `**kwargs`, then you'll get "argument repeated", but the first item will not raise an error.

In short: I'm OK with your version, I'm OK with a version that keeps **kwargs "just in case" (but it won't be useful anymore at least with the current BeautifulSoup.
